feat(tab): introduce fixed variant

asudoh opened the pull request:
The new variant is enabled with `.bx--tabs--fixed` class.

Fixes #1329.

#### Changelog

**New**

- `.bx--tabs--fixed` class, which enables the new tab variant.
- A React prop to enable such class.

#### Testing / Reviewing

Testing should make sure our tabs component is not broken.

asudoh committed:
feat(tab): introduce fixed variant

The new variant is enabled with `.bx--tabs--fixed` class.

Fixes #1329.

asudoh committed:
fix(tabs): fix disabled color for narrow mode

emyarod commented on packages/components/src/components/tabs/tabs.hbs:
```diff
@@ -5,7 +5,7 @@
   LICENSE file in the root directory of this source tree.
 -->
 
```
we may need unique IDs for the examples otherwise the tabs will control the content for all examples

emyarod reviewed (commented):
it looks like the `div` inside flexbox container situation also occurs in our vanilla docs (related https://github.com/carbon-design-system/carbon/pull/4320/commits/621be60ef12208ee712a0cee4fd3c184454011ce)

`.component-example__live--rendered > div` is a `div` inside of a flexbox so it no longer has `width: 100%` by default, and instead it shrinks to fit the content, which looks like this

![image](https://user-images.githubusercontent.com/8265238/67401568-f19b2980-f574-11e9-8154-b304563f89e0.png)

jnm2377 commented on packages/components/src/components/tabs/_tabs.scss:
```diff
@@ -178,6 +207,13 @@
+  .#{$prefix}--tabs--fixed
+    .#{$prefix}--tabs__nav-item:hover:not(.#{$prefix}--tabs__nav-item--disabled) {
+    @include carbon--breakpoint(md) {
+      background-color: $hover-ui;
```
I know this is the token the spec has, but for some reason the color it's receiving is `#e5e5e5` instead of `#cacaca`, so the hover color is lighter and it should be darker. Not sure why the color doesn't match the token though. Is this the right token? @aagonzales 
<img width="739" alt="Screen Shot 2019-10-23 at 11 31 41 AM" src="https://user-images.githubusercontent.com/32556167/67414482-ccb0b180-f588-11e9-9638-9b8820544fee.png">

aagonzales replied:
good catch, it will need to use a different token. Use`$hover-selected-ui` (even though its not "selected" it gets it the correct color) 🤷‍♀

jnm2377 commented:
Just a few things:
- I'm not seeing the tab content background color in the React Storybook.
- Could we move the React example to be it's own story rather than a knob, like our other component examples? I feel like knobs are often hard to find for users/most people don't know to look there.

aagonzales reviewed (changes requested):
**Bugs**
- [ ] height should be `48px / 3rem`
- [x] border top should be `2px`

![image](https://user-images.githubusercontent.com/11670886/67423777-7d737c80-f59a-11e9-8083-5462530a59f9.png)

**A few design tweaks from the original spec we want to change:**
- [x] remove border-bottom from all the tabs
- [x] remove border's to the left and right of the selected tab

![image](https://user-images.githubusercontent.com/11670886/67423997-eeb32f80-f59a-11e9-9278-e0c6ef176f1d.png)

**In the example:**
Can we add a background color to `some-content` area - `$ui-01`

asudoh committed:
fix(tabs): address review comments

asudoh commented:
Thanks @aagonzales @emyarod @jnm2377 for reviewing!

* Removed borders of:
  * Bottom one entierly
  * Left/right ones for selected tab
* Reduced the (top) border size of selected tab
* Changed the background color of selected tab
* Increased the height of the tab for the new variant
* Fixed ID duplication in vanilla
* Styled React tab panel
* Fixed width and top margin of vanilla tab panel
* Made different story for the new variant. Though we have several knobs for different variants (e.g. `light` variant), I see @jnm2377's point wrt difficulty to find out

emyarod reviewed (commented):
the vanilla IDs still seem to be shared since the examples are controlling each other, otherwise looks good to me pending design review

aagonzales reviewed (commented):
Nice! They're looking good, thanks for updating. Just a few bugs still:

- Height of tabs should be `48px/3rem`
- There's a weird delay in state change when the previous `selected` is transitioning to `unselected`
![Oct-24-2019 10-02-25](https://user-images.githubusercontent.com/11670886/67498899-d4825b80-f645-11e9-96bf-093f5fcfe33c.gif)

- On `hover` both borders should disappear, only one is currently 
![Oct-24-2019 10-10-19](https://user-images.githubusercontent.com/11670886/67499354-96d20280-f646-11e9-97dd-389074bffa18.gif)

jnm2377 commented:
hmm looks like the state delay is an existing bug. It happens in the default tabs as well @aagonzales @asudoh

jnm2377 reviewed (approved):
Other than the hover border, LGTM. I would be ok with opening a separate issue/PR for the selected state delay since it's technically a different existing bug/out of scope for this PR, but also would be ok if you wanted to include a fix for it in this PR since you're already working on tabs. 👍

asudoh committed:
fix(tabs): address more review comments

asudoh committed:
chore(devenv): disable onClick action logger for Tabs

asudoh commented:
@aagonzales @emyarod Good catches - Ensured 3rem height, removed right border for selected tab, and resolved duplicate tab panel target. For the detailed animation wrt selected state, I see a performance issue in the Storybook setup. Fixed it that seems to solve or at least alleviate the problem. If the problem still persists I think it's a good idea to defer it, as @jnm2377 saw that it's not a new issue.

aagonzales reviewed (approved):
🎉 awesome! ship it!

asudoh committed:
Merge branch 'master' into tabs-fixed
